File: Trab_ED_Samuel_Finkler_Java/src/trab_ed_samuel_finkler_java/Grafo.java

package trab_ed_samuel_finkler_java;

import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Scanner;
import java.util.logging.Level;
import java.util.logging.Logger;

public class Grafo {
    ArrayList<Vertice> lista_vertices;
    int cont_arestas;

    public Grafo() {
        this.lista_vertices = new ArrayList<Vertice>();
    }

    public ArrayList<Vertice> getLista_vertice() {
        return lista_vertices;
    }

    public void setLista_vertice(ArrayList<Vertice> lista_vertice) {
        this.lista_vertices = lista_vertice;
    }

    public int getCont_arestas() {
        return cont_arestas;
    }

    public void setCont_arestas(int cont_arestas) {
        this.cont_arestas = cont_arestas;
    }
    
    public int grafo_qtd_vertices(){
       return this.lista_vertices.size();
    }
    
    public ArrayList<Vertice> grafo_busca_vertices_saida(Vertice vertice){
        int num_arestas = 0, i=0, j=0;
        ArrayList<Vertice> lista_vertices_saida = new ArrayList<Vertice>();
        for(i=0; i<lista_vertices.size(); i++){
            if(lista_vertices.get(i)==vertice){
                ArrayList<Aresta> lista_arestas = this.lista_vertices.get(i).getLista_arestas();
                for(j=0; j<lista_arestas.size(); j++){
                    lista_vertices_saida.add(lista_arestas.get(j).getVertice_saida());
                    num_arestas++;
                }
            }
        }
        setCont_arestas(num_arestas);
        return lista_vertices_saida;
    }
    
    public ArrayList<Vertice> grafo_busca_vertices_entrada(Vertice vertice){
        int num_arestas = 0, i=0, j=0;
        ArrayList<Vertice> lista_vertices_entrada = new ArrayList<Vertice>();
        for(i=0; i<lista_vertices.size(); i++){
            ArrayList<Aresta> lista_arestas = this.lista_vertices.get(i).getLista_arestas();
            for(j=0; j<lista_arestas.size(); j++){
                if(lista_arestas.get(j).vertice_saida==vertice){
                    System.out.println("entrei no IF");
                    lista_vertices_entrada.add(lista_vertices.get(i));
                    num_arestas++;
                }
            }
        }
        setCont_arestas(num_arestas);
        return lista_vertices_entrada;
    }
    
    public Vertice grafo_busca_vertice_fonte(){
        int entradas=0;
        for(int i=0; i<lista_vertices.size(); i++){
            grafo_busca_vertices_entrada(lista_vertices.get(i));
            entradas = getCont_arestas();
            if(entradas==0){
                return lista_vertices.get(i);
            }
        }
        return null;
    }
    
    public Vertice grafo_busca_vertice_sumidouro(){
        int saidas=0;
        for(int i=0; i<lista_vertices.size(); i++){
            grafo_busca_vertices_saida(lista_vertices.get(i));
            saidas = getCont_arestas();
            if(saidas==0){
                return lista_vertices.get(i);
            }
        }
        return null;
    }
    
    public void grafo_resetar_arestas(){
        int i=0, j=0;
        for(i=0; i<lista_vertices.size()-1; i++){ 
            ArrayList<Aresta> lista_arestas = this.lista_vertices.get(i).getLista_arestas();
            for(j=0; j<lista_arestas.size(); j++){
                lista_arestas.get(j).setValor_ida(0);
                lista_arestas.get(j).setValor_volta(Integer.MAX_VALUE);
                lista_arestas.get(j).setVerificacao_caminho(-1);
            }
        }  
    }
    
    public Aresta grafo_busca_aresta(Vertice origem, Vertice destino){
        return origem.busca_aresta(destino);
    }
    
    public void grafo_remove_aresta(Vertice origem, Vertice destino){
        origem.remove_aresta(destino);
    }
    
    public Vertice grafo_cria_vertice(String nome){
        Vertice vertice = new Vertice(nome);
        this.lista_vertices.add(vertice);
        return vertice;
    }
    
    public Vertice grafo_retorna_vertice_por_nome(String nome){
        int i=0;
        for(i=0; i<lista_vertices.size(); i++){
            if(lista_vertices.get(i).nome.equals(nome)){
                return lista_vertices.get(i);
            }
        }
        return null;
    }
    
    public void grafo_imprime(Vertice verticeFonte){
        int i=0;
        ArrayList<Aresta> listaArestas = verticeFonte.getLista_arestas();
        for(i=0; i<listaArestas.size(); i++){
            System.out.println(""+verticeFonte.nome+" -> "+listaArestas.get(i).getVertice_saida().getNome()+" ("+listaArestas.get(i).getPeso()+")");
            grafo_imprime(listaArestas.get(i).getVertice_saida());
        }
    }
    
    public void grafo_imprime_maior_caminho(Vertice verticeFonte){
        int i=0;
        ArrayList<Aresta> listaArestas = verticeFonte.getLista_arestas();
        for(i=0; i<listaArestas.size(); i++){
            if(listaArestas.get(i).verificacao_caminho==0){
                System.out.println(""+verticeFonte.nome+" -> "+listaArestas.get(i).getVertice_saida().getNome()+" ("+listaArestas.get(i).getPeso()+")");
                grafo_imprime_maior_caminho(listaArestas.get(i).getVertice_saida());
            }
        }
    }
    
    public int grafo_maior_caminho_volta(Vertice verticeFonte, Vertice verticeSumidouro, int idaCaminho){
        
        int voltaCaminho=0, maiorCaminho=0, cont=0, i=0;
        ArrayList<Vertice> v_dest = grafo_busca_vertices_saida(verticeFonte);
        cont = cont_arestas;
        Aresta aresta = null;
        
        if(verticeFonte==verticeSumidouro){
            return idaCaminho;
        }
        
        for(i=0; i<cont; i++){
            voltaCaminho = grafo_maior_caminho_volta(v_dest.get(i), verticeSumidouro, idaCaminho);
            aresta = grafo_busca_aresta(verticeFonte, v_dest.get(i));
            if(voltaCaminho < aresta.getValor_volta()){
                aresta.setValor_volta(voltaCaminho);
                aresta.setVerificacao_caminho(aresta.getValor_ida()-aresta.getValor_volta());
            }
            if(aresta.getVerificacao_caminho()==0){
                maiorCaminho=voltaCaminho-aresta.getPeso();
            }
        }
        return maiorCaminho;
    }
    
    public int grafo_maior_caminho_ida(Vertice verticeFonte, Vertice verticeSumidouro, int idaCaminho){
        int voltaCaminho=0, maiorCaminho=0, cont=0, i=0;
        ArrayList<Vertice> v_dest = grafo_busca_vertices_saida(verticeFonte);
        cont = cont_arestas;
        Aresta aresta = null;
        if(verticeFonte==verticeSumidouro){
            return idaCaminho;
        }
        for(i=0; i<cont; i++){
            aresta = grafo_busca_aresta(verticeFonte, v_dest.get(i));
            if(aresta.getValor_ida() < idaCaminho+aresta.getPeso()){
                aresta.setValor_ida(idaCaminho+aresta.getPeso());
            }
            voltaCaminho = grafo_maior_caminho_ida(v_dest.get(i), verticeSumidouro, aresta.getValor_ida());
            if(voltaCaminho>maiorCaminho){
                maiorCaminho=voltaCaminho;
            }
        }
        return maiorCaminho;
    }
    
    public void imprimir_itens_alcance(Vertice vertice){
        int i=0;
        ArrayList<Item> itens_alcance = vertice.getItens_saida();
        for(i=0; i<itens_alcance.size(); i++){
            System.out.println(""+itens_alcance.get(i).getVertice().getNome());
        }
    }
    
    public void dfs(Vertice vertice, Vertice destino){
        int cont=0, i=0;
        ArrayList<Vertice> v_dest = grafo_busca_vertices_saida(destino);
        cont = cont_arestas;
        
        for(i=0; i<cont; i++){
            dfs(vertice, v_dest.get(i));
            if(vertice.busca_item(v_dest.get(i))==null){
                vertice.adiciona_item(v_dest.get(i));
            }       
        }
    }
    
    private int grafo_verifica_alcance(Vertice v1, Vertice v2){
        int i=0;
        if(v1==v2){
            return 1;
        }
        ArrayList<Item> lista_itens = v1.getItens_saida();
        for(i=0; i<lista_itens.size(); i++){
            if(lista_itens.get(i).vertice == v2  || lista_itens.get(i).vertice == v1){
                return 1;
            }  
        }
        lista_itens = v2.getItens_saida();
        for(i=0; i<lista_itens.size(); i++){
            if(lista_itens.get(i).vertice == v1  || lista_itens.get(i).vertice == v2){
                return 1;
            }
        }
        return 0;
    }
    
    public void grafo_teste_insercao_custo0(Vertice verticeFonte, Vertice verticeSumidouro){
        int i=0, j=0;
        int caminhoCritico = grafo_maior_caminho_ida(verticeFonte, verticeSumidouro, 0);
        grafo_maior_caminho_volta(verticeFonte, verticeSumidouro, caminhoCritico);
        grafo_resetar_arestas();
        int testeCaminhoCritico=0, testealcance=0;
        Item item = null;
        
        for(i=0; i<this.lista_vertices.size(); i++){
            dfs(this.lista_vertices.get(i), this.lista_vertices.get(i));
        }
        
        //Linhas comentadas a seguir são para imprimir a lista de alcance de cada vértice
        /*System.out.println("\nLista de alcançe:");
        for(i=0; i<this.lista_vertices.size(); i++){
            System.out.println("Vertice "+lista_vertices.get(i).getNome());
            this.imprimir_itens_alcance(lista_vertices.get(i));
        }*/
        
        for(i=0; i<this.lista_vertices.size();i++){
            for(j=0; j<this.lista_vertices.size(); j++){
                testealcance = grafo_verifica_alcance(this.lista_vertices.get(i), this.lista_vertices.get(j));
                if(testealcance==0){
                    this.lista_vertices.get(i).insere_aresta(this.lista_vertices.get(j), 0);
                    testeCaminhoCritico = grafo_maior_caminho_ida(verticeFonte, verticeSumidouro, 0);
                    grafo_maior_caminho_volta(verticeFonte, verticeSumidouro, testeCaminhoCritico);
                    if(testeCaminhoCritico>caminhoCritico){
                        System.out.println("\nNovo caminho crítico com custo 0 inserido entre "+this.lista_vertices.get(i).getNome()+" e "+this.lista_vertices.get(j).getNome());
                        System.out.println("Novo caminho crítico:"+testeCaminhoCritico);
                        grafo_imprime_maior_caminho(verticeFonte);
                    }
                    grafo_resetar_arestas();
                    grafo_remove_aresta(this.lista_vertices.get(i), this.lista_vertices.get(j));
                }
            }
        }
    }
    
    void ler_arquivo(){
        Vertice verticeOrigem;
        Vertice verticeDestino;
        String origem, destino;
        int peso, cont=0;
        try {
            Scanner scanner = new Scanner(new FileReader("Rede.txt")).useDelimiter("\\,|\\n");
            while (scanner.hasNext()) {
                cont++;
                origem = scanner.next();
                destino = scanner.next();
                peso = scanner.nextInt();
                
                verticeOrigem = grafo_retorna_vertice_por_nome(origem);
                if(verticeOrigem==null){
                    verticeOrigem = grafo_cria_vertice(origem);
                }
                verticeDestino = grafo_retorna_vertice_por_nome(destino);
                if(verticeDestino==null){
                    verticeDestino = grafo_cria_vertice(destino);
                }
                verticeOrigem.insere_aresta(verticeDestino, peso);
            }
        } catch (FileNotFoundException ex) {
            Logger.getLogger(Grafo.class.getName()).log(Level.SEVERE, null, ex);
        }   
    }
    
}

File: Trab_ED_Samuel_Finkler/grafo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include "grafo.h"

struct aresta{
    int peso;
    int valor_ida;
    int valor_volta;
    int verificacao_caminho;  //se for igual a zero depois do teste faz parte do caminho crítico
    Vertice * vertice_saida;
    struct aresta * prox_aresta;
};

struct vertice{
    char nome[50];
    Item * primeiro_item;  //Vértices alcançaveis pelo vértice atual
    Vertice * prox_vertice;
    struct aresta * primeira_aresta;
    struct aresta * ultima_aresta;
};

struct grafo{
    Vertice * primeiro_vertice;
    Vertice * ultimo_vertice;
};

//Estrutura para guardar todos os vértices que podem ser alcançados por um dado vértice
struct item{
    Vertice* vertice;
    struct item  * prox_item;
};

//funço para criação de um grafo
Grafo * grafo_cria(){
    Grafo * grafo = (Grafo *) malloc(sizeof(Grafo));
    grafo->primeiro_vertice = NULL;
    grafo->ultimo_vertice = NULL;
    return grafo;
}

//Função para liberar a memória ocupada pelo grafo
void grafo_libera (Grafo * grafo){
  int qtVertices= grafo_qtd_vertices(grafo);
  Vertice* vertice_atual;
  Vertice* vertice_aux;
  Aresta* aresta_atual;
  Aresta* aresta_aux;
  vertice_atual = grafo->primeiro_vertice;
  for(int i=0; i<qtVertices; i++){
    aresta_atual = vertice_atual->primeira_aresta;
    while(aresta_atual!=NULL){
      aresta_aux = aresta_atual->prox_aresta;
      free(aresta_atual);
      aresta_atual=aresta_aux;
    }
    vertice_aux = vertice_atual->prox_vertice;
    free(vertice_atual);
    vertice_atual=vertice_aux;
  }
  free(grafo);
}

//Função para criar e retornar ítem
Item * adiciona_item(Item* item, Vertice * vertice){
    Item * novo_item = (Item *)malloc(sizeof(Item));
    novo_item->vertice = vertice;
    novo_item->prox_item = item;
    return novo_item;
}

//Função para buscar um ítem da lista de alcance
Item * busca_item(Item * item, Vertice * vertice){
    while (item!=NULL){
        if (item->vertice == vertice)
            return item;
        item = item->prox_item;
    }
    return NULL;
}

//Função de busca em profundidade alterada para mapear vértices alcançaveis por um dado vértice
static Item* dfs(Grafo * grafo, Vertice * vertice, Item* item){
  int cont = 0;
  Vertice ** v_dest = grafo_busca_vertices_saida(grafo, vertice, &cont);

  for(int i=0;i<cont;i++){
    item = dfs(grafo, v_dest[i], item);
    if(busca_item(item, v_dest[i])==NULL){
      item = adiciona_item(item, v_dest[i]);
    }
  }
  return item;
}

//Função para fins de teste, para verificar se o mapeamento de alcançe ficou correto
/*static void imprimir_itens_alcance(Vertice* vertice){
  Item* item = vertice->primeiro_item;
  while(item!=NULL){
    printf("\n%s", item->vertice->nome);
    item= item->prox_item;
  }
}*/

//retorna 1 se um vértice possui alcance até o outro
static int grafo_verifica_alcance(Grafo* grafo, Vertice* v1, Vertice* v2){
  if(v1==v2){
    return 1;
  }

  Item* item = v1->primeiro_item;
  while(item!=NULL){
    if(item->vertice==v2 || item->vertice==v1){
      return 1;
    }
    item = item->prox_item;
  }
  item = v2->primeiro_item;
  while(item!=NULL){
    if(item->vertice==v1 || item->vertice==v2){
      return 1;
    }
    item = item->prox_item;
  }
  return 0;
}

//Função para retornar a quantidade de vértices do grafo
int grafo_qtd_vertices (Grafo * grafo){
    Vertice * vertice = grafo->primeiro_vertice;
    int qtd = 0;
    while (vertice!=NULL){
        qtd++;
        vertice = vertice->prox_vertice;
    }
    return qtd;
}

//Função para retornar vértices de saída de um dado vértice
Vertice ** grafo_busca_vertices_saida(Grafo * grafo, Vertice * vertice, int * cont){
    int num_arestas =0;
    Aresta * aresta = vertice->primeira_aresta;
    while (aresta!=NULL){
        num_arestas++;
        aresta = aresta->prox_aresta;
    }
    Vertice ** vertices = (Vertice **)malloc(sizeof(Vertice * ) * num_arestas);
    int num_aresta = 0;
    aresta = vertice->primeira_aresta;
    while (aresta!=NULL && num_aresta<num_arestas){
        vertices[num_aresta] = aresta->vertice_saida;
        aresta = aresta->prox_aresta;
        num_aresta++;
    }
    *cont = num_arestas;
    return vertices;
}

//Função para retornar vértices de entrada de um dado vértice
Vertice ** grafo_busca_vertices_entrada(Grafo * grafo, Vertice * vertice, int * cont){
    Vertice* vertice_atual = grafo->primeiro_vertice;
    Aresta* aresta_atual = NULL;
    int contador_arestas_saida=0;

    while(vertice_atual!=NULL){
      aresta_atual=vertice_atual->primeira_aresta;
      while(aresta_atual!=NULL){
        if(aresta_atual->vertice_saida == vertice){
          contador_arestas_saida++;
          break;
        }
        aresta_atual=aresta_atual->prox_aresta;
      }
      vertice_atual=vertice_atual->prox_vertice;
    }

    Vertice ** vertices = (Vertice **)malloc(sizeof(Vertice*) * contador_arestas_saida);

    vertice_atual = grafo->primeiro_vertice;
    int aux_cont=0;
    while(vertice_atual!=NULL && aux_cont<contador_arestas_saida){
      aresta_atual=vertice_atual->primeira_aresta;
      while(aresta_atual!=NULL){
        if(aresta_atual->vertice_saida == vertice){
          vertices[aux_cont] = vertice_atual;
          aux_cont++;
        }
        aresta_atual = aresta_atual->prox_aresta;
      }
      vertice_atual=vertice_atual->prox_vertice;
    }
    *cont = contador_arestas_saida;
    return vertices;
}

//Função para encontrar e retornar o vértice fonte do grafo
Vertice * grafo_busca_vertice_fonte(Grafo * grafo){
  int cont=0;
  Vertice * vertice = grafo->primeiro_vertice;
  while(vertice!=NULL){
    grafo_busca_vertices_entrada(grafo, vertice, &cont);
    if(cont==0){
      return vertice;
    }
    vertice = vertice->prox_vertice;
  }
  return NULL;
}

//Função para encontrar e retornar o vértice sumidouro do grafo
Vertice * grafo_busca_vertice_sumidouro(Grafo * grafo){
  int cont=0;
  Vertice * vertice = grafo->primeiro_vertice;
  while(vertice!=NULL){
    grafo_busca_vertices_saida(grafo, vertice, &cont);
    if(cont==0){
      return vertice;
    }
    vertice = vertice->prox_vertice;
  }
  return NULL;
}


//Função para resetar os valores padrão dos atributos das arestas usados para achar o maior caminho
void grafo_resetar_arestas(Grafo* grafo){
  Vertice* vertice = grafo->primeiro_vertice;
  Aresta* aresta = NULL;
  while(vertice!=NULL){
    aresta = vertice->primeira_aresta;
    while(aresta!=NULL){
      aresta->valor_ida=0;
      aresta->valor_volta=INT_MAX;
      aresta->verificacao_caminho=-1;

      aresta = aresta->prox_aresta;
    }
    vertice=vertice->prox_vertice;
  }
}

//Função para inserir uma aresta entre dois vértices
void grafo_insere_aresta (Grafo * grafo, Vertice * vertice1, Vertice * vertice2, int peso){
    Aresta * aresta = (Aresta*) malloc(sizeof(Aresta));
    aresta->peso = peso;
    aresta->valor_ida= 0;
    aresta->valor_volta = INT_MAX;
    aresta->verificacao_caminho = -1;
    aresta->vertice_saida = vertice2;
    aresta->prox_aresta = NULL;

    if (vertice1->ultima_aresta!=NULL){
        vertice1->ultima_aresta->prox_aresta = aresta;
        vertice1->ultima_aresta = aresta;
    }else{
        vertice1->primeira_aresta = aresta;
        vertice1->ultima_aresta = aresta;
    }
}

//Função para buscar uma aresta entre dois vértices
Aresta * grafo_busca_aresta(Grafo * grafo, Vertice * vertice1, Vertice * vertice2){
    Aresta* aresta_atual = vertice1->primeira_aresta;
    while(aresta_atual!=NULL){
      if(aresta_atual->vertice_saida == vertice2){
        return aresta_atual;
      }
      aresta_atual=aresta_atual->prox_aresta;
    }
    return NULL;
}

//Função para remover uma aresta entre dois vértices
void grafo_remove_aresta (Grafo* grafo, Vertice * vertice1, Vertice * vertice2){
  Aresta* ant = NULL;
  Aresta* aresta = vertice1->primeira_aresta;
  while(aresta!=vertice1->ultima_aresta){
    ant = aresta;
    aresta = aresta->prox_aresta;
  }
  if(ant==NULL){
    vertice1->primeira_aresta = NULL;
    vertice1->ultima_aresta = NULL;
  }else{
    ant->prox_aresta=NULL;
    vertice1->ultima_aresta = ant;
  }
  free(aresta);
}

//Funcão de criação de vértice
Vertice * grafo_cria_vertice(Grafo * grafo, const char nome[]){
    Vertice * v = (Vertice *)malloc(sizeof(Vertice));
    strcpy(v->nome, nome);
    v->primeiro_item = NULL;
    v->prox_vertice = NULL;
    v->primeira_aresta = NULL;
    v->ultima_aresta = NULL;

    if (grafo->ultimo_vertice!=NULL){
        grafo->ultimo_vertice->prox_vertice = v;
        grafo->ultimo_vertice = v;
    }else{
        grafo->primeiro_vertice = v;
        grafo->ultimo_vertice = v;
    }
    return v;
}

Vertice * grafo_retorna_vertice(Grafo * grafo, int indice){
    Vertice * vertice = grafo->primeiro_vertice;
    int qtd = 0;
    while (vertice!=NULL){
        if (qtd == indice)
            return vertice;
        qtd++;
        vertice = vertice->prox_vertice;
    }
    return NULL;
}

//Função para buscar e retornar um vértice de acordo com seu nome
Vertice * grafo_retorna_vertice_por_nome(Grafo * grafo, char* nome){
    Vertice * vertice = grafo->primeiro_vertice;
    while (vertice!=NULL){
        if (!strcmp(nome, vertice->nome))
            return vertice;
        vertice = vertice->prox_vertice;
    }
    return NULL;
}

//Função para retornar nome de um vértice
char * grafo_retorna_nome(Vertice * vertice){
    return vertice->nome;
}

//Função de teste usada para imprimir todos os caminhos do grafo
void grafo_imprime(Grafo * grafo, Vertice * verticeFonte){
  Aresta* aresta = verticeFonte->primeira_aresta;
  while(aresta != NULL){
      printf("\n%s -> %s (%d)", verticeFonte->nome, aresta->vertice_saida->nome, aresta->peso);
      grafo_imprime(grafo, aresta->vertice_saida);
    aresta = aresta->prox_aresta;
  }
}

//Função para imprimir o maior caminho
void grafo_imprime_maior_caminho(Grafo * grafo, Vertice * verticeFonte){
  Aresta* aresta = verticeFonte->primeira_aresta;
  while(aresta != NULL){
    if(aresta->verificacao_caminho==0){
      printf("%s -> %s (%d)  |  ", verticeFonte->nome, aresta->vertice_saida->nome, aresta->peso);
      grafo_imprime_maior_caminho(grafo, aresta->vertice_saida);
    }
    aresta = aresta->prox_aresta;
  }
}

int grafo_maior_caminho_volta(Grafo * grafo, Vertice * verticeFonte, Vertice * verticeSumidouro, int idaCaminho){
  int voltaCaminho=0, maiorCaminho=0;
  int cont = 0;
  Vertice ** v_dest = grafo_busca_vertices_saida(grafo, verticeFonte, &cont);
  Aresta * aresta = NULL;

  if(verticeFonte==verticeSumidouro){
    return idaCaminho;
  }

  for (int i=0;i<cont;i++){
    voltaCaminho = grafo_maior_caminho_volta(grafo, v_dest[i], verticeSumidouro, idaCaminho);
    aresta = grafo_busca_aresta(grafo, verticeFonte, v_dest[i]);

    if(voltaCaminho < aresta->valor_volta){
      aresta->valor_volta = voltaCaminho;
      aresta->verificacao_caminho = aresta->valor_ida-aresta->valor_volta;
    }
    if(aresta->verificacao_caminho == 0){
      maiorCaminho=voltaCaminho-aresta->peso;
    }
  }
  return maiorCaminho;
}

int grafo_maior_caminho_ida(Grafo * grafo, Vertice * verticeFonte, Vertice * verticeSumidouro, int idaCaminho){
  int voltaCaminho=0, maiorCaminho=0;
  int cont = 0;
  Vertice ** v_dest = grafo_busca_vertices_saida(grafo, verticeFonte, &cont);
  Aresta * aresta = NULL;

  if(verticeFonte==verticeSumidouro){
    return idaCaminho;
  }
  for (int i=0;i<cont;i++){
    aresta = grafo_busca_aresta(grafo, verticeFonte, v_dest[i]);
    if(aresta->valor_ida < idaCaminho+aresta->peso){
      aresta->valor_ida = idaCaminho+aresta->peso;
    }
    voltaCaminho = grafo_maior_caminho_ida(grafo, v_dest[i], verticeSumidouro, aresta->valor_ida);
    if(voltaCaminho > maiorCaminho){
      maiorCaminho = voltaCaminho;
    }
  }
  return maiorCaminho;
}

void grafo_teste_insercao_custo0(Grafo* grafo, Vertice* verticeFonte, Vertice* verticeSumidouro){
  int caminhoCritico = grafo_maior_caminho_ida(grafo, verticeFonte, verticeSumidouro, 0);
  grafo_maior_caminho_volta(grafo, verticeFonte, verticeSumidouro, caminhoCritico);
  grafo_resetar_arestas(grafo);
  int testeCaminhoCritico = 0, testeAlcance=0;
  Vertice* v1=verticeFonte;
  Vertice* v2=NULL;
  Item* item = NULL;

  while(v1!=NULL){
    item = v1->primeiro_item;
    v1->primeiro_item = dfs(grafo, v1, item);
    v1 = v1->prox_vertice;
  }
  v1=verticeFonte;

  /*printf("\n\nLISTA DE ALCANÇE");
  while(v1!=NULL){
    printf("\nAo alcance de %s", v1->nome);
    imprimir_itens_alcance(v1);
    v1 = v1->prox_vertice;
  }
  v1=verticeFonte;*/

  printf("\n\n");
  while(v1!=NULL){
    v2 = verticeFonte;
    while(v2!=NULL){
      testeAlcance = grafo_verifica_alcance(grafo, v1, v2);
      if(testeAlcance==0){
        grafo_insere_aresta(grafo, v1, v2, 0);
        testeCaminhoCritico = grafo_maior_caminho_ida(grafo, verticeFonte, verticeSumidouro, 0);
        grafo_maior_caminho_volta(grafo, verticeFonte, verticeSumidouro, testeCaminhoCritico);
        if(testeCaminhoCritico>caminhoCritico){
          printf("\nNovo caminho cŕitico com custo 0 inserido entre %s e %s", grafo_retorna_nome(v1), grafo_retorna_nome(v2));
          printf("\nNovo caminho crítico: %d", testeCaminhoCritico);
          printf("\n");
          grafo_imprime_maior_caminho(grafo, verticeFonte);
          printf("\n");
        }
        grafo_resetar_arestas(grafo);
        grafo_remove_aresta(grafo, v1, v2);
      }
      v2 = v2->prox_vertice;
    }
    v1 = v1->prox_vertice;
  }
}

//Função para ler arquivo, inserir arestas e vértices no grafo
void lerArquivo(Grafo* grafo){
  Vertice* verticeOrigem = NULL;
  Vertice* verticeDestino = NULL;
  char tarefaOrigem[50], tarefaDestino[50];
  int custo;

  FILE *file;
  file = fopen("Rede.txt", "r");

  if (file == NULL){
    printf("\nERRO! O arquivo não foi aberto!\n");
    exit(1);
  }else{
     printf("\nO arquivo foi aberto com sucesso!");
  }

  while(fscanf(file,"\n%50[^,] ,%50[^,] ,%d", tarefaOrigem, tarefaDestino, &custo)!=EOF){
      verticeOrigem = grafo_retorna_vertice_por_nome(grafo, tarefaOrigem);
      if(verticeOrigem==NULL){
        verticeOrigem = grafo_cria_vertice(grafo, tarefaOrigem);
      }
      verticeDestino = grafo_retorna_vertice_por_nome(grafo, tarefaDestino);
      if(verticeDestino==NULL){
        verticeDestino = grafo_cria_vertice(grafo, tarefaDestino);
      }
      grafo_insere_aresta(grafo, verticeOrigem, verticeDestino, custo);
      printf("\n%s,%s,%d", tarefaOrigem, tarefaDestino, custo);
  }

  fclose(file);
}
